### PR TITLE
Support parallel deployment of etcd operator and Cilium

### LINKED ISSUE #4314: etcd-operator compatible Cilium bootstrap
## Goal

Allow to deploy Cilium with a etcd-operator as part of the deployment resources and have Cilium provide networking connectivity first to allow etcd-operator to be deployed. Then have Cilium connect to the deployed etcd and establish the identity system for global policy. Allow policies to be applied to etcd using fixed/init policies without requiring kvstore access.

## Tasks
- [x] Reserve numeric identity 128-256 for fixed numeric identities
  - [x] Introduce mapping of reserved:<name>
- [x] Add a special label `io.cilium.fixed-identity=identity-name` to allow containers to select a fixed identity
- [ ] Make kvstore connect async on bootstrap with a 3 minute hard timeout. cilium status should not report unhealthy state in this 3 minute period if kvstore cannot be reached. As soon as connectivity has been established once, loss of connectivity to all cluster members is a unhealthy state again.
- [ ] Add desired/realized state logic to endpoint when identity is being resolved for endpoint. Export this via API for visibility.
- [ ] Verify that k8s service handler is fine while kvstore is not reachable yet. We might have to defer allocation of service IDs to later.

## Explanation
- etcd-operator can be defined with special label to select fixed identity
- Respective endpoint will come up immediately without requiring kvstore access.
- All other endpoints will be in init identity state and will have identity resolution pending
- As kvstore connectivity is achieved, other endpoints will transition to real identities

### PR DESCRIPTION
**Pull request by aanm**
**Summary of changes**:

This PR enables Cilium without a kvstore connection for 3 minutes which might give enough time for a cluster to set up an etcd-operator.

Fixes: #4314 

```release-note
Add support for etcd-operator to ease Cilium deployment.
```

Will require documentation changes. 
TL;DR docs:
```
# create etcd-operator
kubectl create -f examples/kubernetes/addons/etcd-operator
# add arguments to cilium.yaml
--fixed-identity-mapping=128=kv-store
--fixed-identity-mapping=129=kube-dns
# point etcd configuration in cilium configmap to
---
endpoints:
- http://cilium-etcd.kube-system.svc.cluster.local:2379
# add label to kube-dns pod
kubectl label -n kube-system pod kube-dns-7dcc557ddd-k4jz6 io.cilium.fixed-identity=kube-dns
# install cilium
kubectl create -f cilium.yaml
# wait a couple minutes until all endpoints are up and running with a sec ID
# remove label from kube-dns pod
kubectl label -n kube-system pod kube-dns-7dcc557ddd-k4jz6 io.cilium.fixed-identity-
# done
```

Depends on #4797  and #4775

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4826)

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-missed-k8s

**Comment by aanm** [comment]
previous missed-k8s were hit by https://github.com/cilium/cilium/issues/4851

**Comment by aanm** [comment]
previous missed-k8s were hit by https://github.com/cilium/cilium/issues/4852

**Comment by aanm** [comment]
test-missed-k8s

**Comment by aanm** [comment]
test-me-please

File: daemon/state.go [hunk 1]
```diff
@@ -161,12 +170,18 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState) {
...
-				epRegenerated <- false
-				return
+
+			if ep.SecurityIdentity != nil {
+				if oldSecID := ep.SecurityIdentity.ID; identity.ID != oldSecID {
+					log.WithFields(logrus.Fields{
+						logfields.EndpointID:              ep.ID,
+						logfields.IdentityLabels + ".old": oldSecID,
+						logfields.IdentityLabels + ".new": identity.ID,
+					}).Info("Security label ID for endpoint is different that the one stored, updating")
```
Comment (ianvernon) [inline daemon/state.go]: To ensure consistent vocabulary: security label ID --> security identity

is different that the one stored --> is different from the security identity restored for the endpoint

File: pkg/kvstore/etcd.go [hunk 2]
```diff
@@ -229,39 +239,59 @@ func newEtcdClient(config *client.Config, cfgPath string) (BackendOperations, er
...
+		}
+		if err != nil {
+			log.Fatal(err)
+			return
+		}
+
+		log.Debugf("Session received")
+		s = session
+		// Run renewLease once the firstSession is received
+		if err := ec.renewLease(); err != nil {
```
Comment (ianvernon) [inline pkg/kvstore/etcd.go:285]: Why do we need to renew the lease once the first session is received?

File: pkg/node/local_node.go [hunk 3]
```diff
@@ -46,5 +46,10 @@ func ConfigureLocalNode() error {
 
 	UpdateNode(Identity{Name: localNode.Name}, &localNode, TunnelRoute, nil)
 
-	return registerNode()
+	go func() {
+		if err := registerNode(); err != nil {
```
Comment (tgraf) [inline pkg/node/local_node.go]: this needs a timeout and then failure

File: daemon/state.go [hunk 4]
```diff
@@ -149,9 +149,18 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState) {
...
 		go func(ep *endpoint.Endpoint, epRegenerated chan<- bool) {
-			ep.Mutex.Lock()
-
+			ep.Mutex.RLock()
 			scopedLog := log.WithField(logfields.EndpointID, ep.ID)
+			// Filter the restored labels with the new daemon's filter
+			l, _ := labels.FilterLabels(ep.OpLabels.IdentityLabels())
+			ep.Mutex.RUnlock()
+
+			identity, _, err := identityPkg.AllocateIdentity(l)
```
Comment (tgraf) [inline daemon/state.go:158]: Where is the actual change in this commit? This was already in the background.

File: pkg/kvstore/allocator/allocator.go [hunk 5]
```diff
@@ -282,11 +282,11 @@ func NewAllocator(basePath string, typ AllocatorKey, opts ...AllocatorOption) (*
 		return nil, errors.New("Maximum ID must be greater than minimum ID")
 	}
 
-	if err := a.startWatchAndWait(); err != nil {
-		return nil, err
-	}
+	go func() {
+		a.startWatchAndWait()
```
Comment (tgraf) [inline pkg/kvstore/allocator/allocator.go]: Please catch the error and error out with fatal otherwise we may continue without watching for identity changes

**Review by tgraf — CHANGES_REQUESTED** [review]

File: pkg/node/local_node.go (hunk 3 above)
Comment (aanm) [reply pkg/node/local_node.go]: Done

File: daemon/state.go (hunk 4 above)
Comment (aanm) [reply daemon/state.go:158]: The endpoint lock was being held while doing a new identity allocation, which is itself a blocking function. This would block `cilium endpoint list` as we couldn't retrieve the list of endpoints since they would all be blocked while waiting for a new identity..

File: daemon/state.go (hunk 1 above)
Comment (aanm) [reply daemon/state.go]: Done

File: pkg/kvstore/allocator/allocator.go (hunk 5 above)
Comment (aanm) [reply pkg/kvstore/allocator/allocator.go]: Done

**Comment by aanm** [comment]
test-me-please

**Comment by tgraf** [comment]
NACK found

**Comment by tgraf** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]
